lambdas/pkgselect/index.py

```diff
@@ -88,14 +104,46 @@ def lambda_handler(request):
     """
     bucket = request.args['bucket']
     key = request.args['manifest']
-    aws_access_key_id = request.args['access_key']
-    aws_secret_access_key = request.args['secret_key']
-    aws_session_token = request.args['session_token']
     prefix = request.args.get('prefix')
     logical_key = request.args.get('logical_key')
+    access_key = request.args.get('access_key')
+    secret_key = request.args.get('secret_key')
+    session_token = request.args.get('session_token')
+    allow_anonymous_access = bool(os.getenv('ALLOW_ANONYMOUS_ACCESS'))
+
+    # If credentials are passed in, use them
+    # for the client. If no credentials are supplied, test that
+    # the manifest object is publicly accessible. If so, create
+    # an s3 client using the underlying IAM role's permissions.
+
+    if access_key and secret_key and session_token:
+        s3_client = create_s3_client(
+            aws_access_key_id=access_key,
+            aws_secret_access_key=secret_key,
+            aws_session_token=session_token
+        )
+    elif (
+        allow_anonymous_access and
+        access_key is None and
+        secret_key is None and
+        session_token is None
+    ):
+        # Test to see if the target key is publicly accessible. If not, the call
+        # below will raise and exception and return a 403 response
+        anons3 = boto3.client('s3', config=Config(signature_version=UNSIGNED))
+        response = anons3.head_object(Bucket=bucket, Key=key)
```

Comment:
this assignment is never used, so don't assign

Reply:
it's much cleaner to try/except here and specifically deal with 403s. i know if you hit an error that the whole function will bomb but what if it's just a throttling error and not permissions related?